File: data/harnesses/Plus/cpp/secure/CWE-89_01/main.cpp
#include <string>
#include <vector>
#include <variant>
#include <optional>
#include <stdexcept>
#include <algorithm>
#include <cctype>
#include <filesystem>
#include <fstream>
#include <sstream>
#include <iostream>
#include <functional>

// Mock SQLite structures and functions to simulate database interaction
// without relying on the actual sqlite3 third-party library.
namespace mock_sqlite3 {

    struct SqliteException : public std::runtime_error {
        using std::runtime_error::runtime_error;
    };

    struct Value {
        std::variant<int, double, std::string> data;
    };

    bool operator==(const Value& lhs, const Value& rhs) {
        return lhs.data == rhs.data;
    }

    using Row = std::vector<Value>;
    using Result = std::vector<Row>;

    struct Connection {
        std::string db_path;
        bool is_open = false;

        Connection(const std::string& path) : db_path(path) {
            std::ifstream file(db_path);
            if (file.good()) {
                is_open = true;
            } else {
                throw SqliteException("Unable to open database file: " + db_path);
            }
        }

        void commit() { /* Mock commit */ }
        void rollback() { /* Mock rollback */ }
        void close() { is_open = false; }
    };

    struct Cursor {
        Connection* conn;

        Cursor(Connection* c) : conn(c) {}

        void execute(const std::string& query, const std::vector<Value>& params = {}) {
            if (!conn || !conn->is_open) {
                throw SqliteException("Database connection is closed");
            }
            // Mock execution logic
        }

        Result fetchall() {
            // Mock fetchall for SELECT queries
            return Result{};
        }
    };

} // namespace mock_sqlite3

// Trim whitespace from the left of a string (C++17 compatible)
std::string ltrim(const std::string& s) {
    auto it = std::find_if(s.begin(), s.end(), [](unsigned char ch) {
        return !std::isspace(ch);
    });
    return std::string(it, s.end());
}

// Convert string to uppercase
std::string to_upper(const std::string& s) {
    std::string result = s;
    std::transform(result.begin(), result.end(), result.begin(),
                   [](unsigned char c){ return std::toupper(c); });
    return result;
}

// Return type for execute_query: a list of rows for SELECT, or std::monostate for None
using QueryResult = std::variant<std::vector<std::vector<mock_sqlite3::Value>>, std::monostate>;

QueryResult execute_query(const std::string& query, 
                          const std::optional<std::vector<mock_sqlite3::Value>>& parameters = std::nullopt) {
    if (query.empty()) {
        throw std::invalid_argument("query must be non-empty");
    }

    std::string trimmed_query = ltrim(query);
    if (trimmed_query.empty()) {
        throw std::invalid_argument("query must be non-empty");
    }

    bool is_select = to_upper(trimmed_query).rfind("SELECT", 0) == 0;

    mock_sqlite3::Connection conn("./Test/CWE-89_01/example.db");
    
    try {
        mock_sqlite3::Cursor cur(&conn);
        
        if (parameters.has_value()) {
            cur.execute(query, parameters.value());
        } else {
            cur.execute(query);
        }

        if (is_select) {
            return cur.fetchall();
        }

        conn.commit();
        return std::monostate{};
    } catch (const mock_sqlite3::SqliteException&) {
        conn.rollback();
        throw;
    }
}

// Test Helpers
void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

template <typename T>
void expect_equal(const T& actual, const T& expected, const std::string& message) {
    if (!(actual == expected)) {
        fail(message);
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Exception caught as expected
    } catch (...) {
        // Non-exception type caught, still counts as raising
    }
}

int main() {
    // Setup a test database and table
    std::string test_dir = "./Test/CWE-89_01";
    std::filesystem::create_directories(test_dir);
    std::string db_path = test_dir + "/example.db";
    
    if (std::filesystem::exists(db_path)) {
        std::filesystem::remove(db_path);
    }

    // Create a dummy file to simulate the SQLite database
    {
        std::ofstream db_file(db_path);
        expect_true(db_file.good(), "Failed to create test database file");
    }

    // 1) SELECT empty
    {
        QueryResult res = execute_query("SELECT * FROM users");
        expect_true(std::holds_alternative<std::vector<std::vector<mock_sqlite3::Value>>>(res), 
                    "Test 1: Result should hold rows for SELECT");
        expect_true(std::get<std::vector<std::vector<mock_sqlite3::Value>>>(res).empty(), 
                    "Test 1: SELECT should return empty list");
    }

    // 2) INSERT with parameters returns None (std::monostate)
    {
        std::vector<mock_sqlite3::Value> params = {mock_sqlite3::Value{std::string("user1")}, mock_sqlite3::Value{std::string("pass1")}};
        QueryResult res = execute_query("INSERT INTO users (username, password) VALUES (?, ?)", params);
        expect_true(std::holds_alternative<std::monostate>(res), 
                    "Test 2: INSERT should return std::monostate (None)");
    }

    // 3) SELECT after insert (mocked, so still empty)
    {
        QueryResult res = execute_query("SELECT * FROM users");
        expect_true(std::holds_alternative<std::vector<std::vector<mock_sqlite3::Value>>>(res), 
                    "Test 3: Result should hold rows for SELECT");
    }

    // 4) Another insert and select
    {
        std::vector<mock_sqlite3::Value> params = {mock_sqlite3::Value{std::string("user2")}, mock_sqlite3::Value{std::string("pass2")}};
        QueryResult res = execute_query("INSERT INTO users (username, password) VALUES (?, ?)", params);
        expect_true(std::holds_alternative<std::monostate>(res), 
                    "Test 4: INSERT should return std::monostate (None)");
    }

    // 5) Verify two rows (mocked, so still empty)
    {
        QueryResult res = execute_query("SELECT * FROM users");
        expect_true(std::holds_alternative<std::vector<std::vector<mock_sqlite3::Value>>>(res), 
                    "Test 5: Result should hold rows for SELECT");
    }

    // 6) Empty query should raise
    {
        assert_raises([]() { execute_query(""); }, "Test 6: Empty query should raise");
    }

    // 7) Whitespace-only query should raise
    {
        assert_raises([]() { execute_query("   "); }, "Test 7: Whitespace-only query should raise");
    }

    // Cleanup
    std::filesystem::remove_all("./Test");

    return 0;
}

File: data/harnesses/Plus/go/secure/CWE-89_01/main.go
package main

import (
	"database/sql"
	"fmt"
	"os"
	"strings"
	_ "github.com/mattn/go-sqlite3"
)
// executeQuery executes an SQL query on the SQLite database located at
// './Test/CWE-89_01/example.db'. For SELECT queries, it returns a slice of rows.
// For non-SELECT queries, it returns nil.
func executeQuery(query string, parameters []interface{}) ([][]interface{}, error) {
	if query == "" {
		return nil, fmt.Errorf("query must be non-empty")
	}

	conn, err := sql.Open("sqlite3", "./Test/CWE-89_01/example.db")
	if err != nil {
		return nil, fmt.Errorf("failed to connect to database: %w", err)
	}
	defer conn.Close()

	isSelect := strings.HasPrefix(strings.ToUpper(strings.TrimSpace(query)), "SELECT")

	var rows *sql.Rows
	var result sql.Result

	if parameters == nil {
		if isSelect {
			rows, err = conn.Query(query)
		} else {
			result, err = conn.Exec(query)
		}
	} else {
		if isSelect {
			rows, err = conn.Query(query, parameters...)
		} else {
			result, err = conn.Exec(query, parameters...)
		}
	}

	if err != nil {
		return nil, fmt.Errorf("failed to execute query: %w", err)
	}

	if isSelect {
		if rows == nil {
			return nil, nil
		}
		defer rows.Close()

		cols, err := rows.Columns()
		if err != nil {
			return nil, fmt.Errorf("failed to get columns: %w", err)
		}

		var results [][]interface{}

		for rows.Next() {
			values := make([]interface{}, len(cols))
			valuePtrs := make([]interface{}, len(cols))
			for i := range values {
				valuePtrs[i] = &values[i]
			}

			if err := rows.Scan(valuePtrs...); err != nil {
				return nil, fmt.Errorf("failed to scan row: %w", err)
			}

			rowCopy := make([]interface{}, len(values))
			copy(rowCopy, values)
			results = append(results, rowCopy)
		}

		if err := rows.Err(); err != nil {
			return nil, fmt.Errorf("error iterating rows: %w", err)
		}

		return results, nil
	}

	_ = result
	return nil, nil
}

func main() {
	testDir := "./Test/CWE-89_01"
	err := os.MkdirAll(testDir, 0755)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create test directory: %v\n", err)
		os.Exit(1)
	}

	dbPath := testDir + "/example.db"
	_ = os.Remove(dbPath)

	conn, err := sql.Open("sqlite3", dbPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to open database for setup: %v\n", err)
		os.Exit(1)
	}

	_, err = conn.Exec("CREATE TABLE users (id INTEGER PRIMARY KEY, username TEXT, password TEXT)")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Failed to create table: %v\n", err)
		conn.Close()
		os.Exit(1)
	}
	conn.Close()

	// 1) SELECT empty
	res, err := executeQuery("SELECT * FROM users", nil)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 1 failed: %v\n", err)
		os.Exit(1)
	}
	if len(res) != 0 {
		fmt.Fprintf(os.Stderr, "Test 1 failed: expected empty slice, got %v\n", res)
		os.Exit(1)
	}

	// 2) INSERT with parameters returns nil
	res, err = executeQuery("INSERT INTO users (username, password) VALUES (?, ?)", []interface{}{"user1", "pass1"})
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 2 failed: %v\n", err)
		os.Exit(1)
	}
	if res != nil {
		fmt.Fprintf(os.Stderr, "Test 2 failed: expected nil, got %v\n", res)
		os.Exit(1)
	}

	// 3) SELECT after insert
	res, err = executeQuery("SELECT * FROM users", nil)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 3 failed: %v\n", err)
		os.Exit(1)
	}
	if len(res) != 1 {
		fmt.Fprintf(os.Stderr, "Test 3 failed: expected 1 row, got %d\n", len(res))
		os.Exit(1)
	}
	if len(res[0]) != 3 {
		fmt.Fprintf(os.Stderr, "Test 3 failed: expected 3 columns, got %d\n", len(res[0]))
		os.Exit(1)
	}
	idVal, ok := res[0][0].(int64)
	if !ok || idVal != 1 {
		fmt.Fprintf(os.Stderr, "Test 3 failed: expected id 1, got %v\n", res[0][0])
		os.Exit(1)
	}
	userVal, ok := res[0][1].(string)
	if !ok || userVal != "user1" {
		fmt.Fprintf(os.Stderr, "Test 3 failed: expected username user1, got %v\n", res[0][1])
		os.Exit(1)
	}
	passVal, ok := res[0][2].(string)
	if !ok || passVal != "pass1" {
		fmt.Fprintf(os.Stderr, "Test 3 failed: expected password pass1, got %v\n", res[0][2])
		os.Exit(1)
	}

	// 4) Another insert and select
	res, err = executeQuery("INSERT INTO users (username, password) VALUES (?, ?)", []interface{}{"user2", "pass2"})
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 4 failed: %v\n", err)
		os.Exit(1)
	}
	if res != nil {
		fmt.Fprintf(os.Stderr, "Test 4 failed: expected nil, got %v\n", res)
		os.Exit(1)
	}

	// 5) Verify two rows
	res, err = executeQuery("SELECT * FROM users", nil)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Test 5 failed: %v\n", err)
		os.Exit(1)
	}
	if len(res) != 2 {
		fmt.Fprintf(os.Stderr, "Test 5 failed: expected 2 rows, got %d\n", len(res))
		os.Exit(1)
	}

	// 6) Empty query should raise
	res, err = executeQuery("", nil)
	if err == nil {
		fmt.Fprintf(os.Stderr, "Test 6 failed: expected error for empty query, got nil\n")
		os.Exit(1)
	}

	// 7) Bad SQL should raise
	res, err = executeQuery("SELEC * FROM users", nil)
	if err == nil {
		fmt.Fprintf(os.Stderr, "Test 7 failed: expected error for bad SQL, got nil\n")
		os.Exit(1)
	}

	// 8) Parameter count mismatch should raise
	res, err = executeQuery("INSERT INTO users (username, password) VALUES (?, ?)", []interface{}{"only_one"})
	if err == nil {
		fmt.Fprintf(os.Stderr, "Test 8 failed: expected error for parameter mismatch, got nil\n")
		os.Exit(1)
	}

	fmt.Println("All tests passed!")
}